### ISSUE TITLE
AI features roadmap: on-device LLM / BYOK / hosted-credits Chat tiers

### ISSUE THREAD

**Issue by rahatsayyed** [rahatsayyed/raqm#3]
# Raqm AI Features — Roadmap

Short planning doc for Raqm's AI feature set. Full per-tier specs come later,
once each tier's spike/design work is done. Order below is the build order.

## Guiding principle (applies to every tier)

The AI must never be the source of truth for a number. A model only
reasons or phrases an answer from data a deterministic query already
retrieved from SQLite. This keeps every tier grounded and stops the AI
from stating a spending figure it never actually looked up.

## Tier 1 — On-device AI (free)

**Spike first, in a new throwaway app — not in Raqm.**

- Build a small, separate test app with mock transaction data. Do not
  touch Raqm's codebase until this proves a model works.
- Test multiple on-device candidates: `BitNet b1.58`, `MobileMoE`, and
  any other real option found during setup research.
- `BitNet b1.58` risk note (2026-09-14 research): `bitnet.cpp` has no
  official Android/NDK build support today, one known Android (Termux)
  compile attempt failed with no fix, and no on-phone tokens/sec have
  been published anywhere. No JNI wrapper exists (unlike `llama.cpp`,
  which has an official `llama.android` project). The real GGUF file
  is 1.19 GB, not the ~400 MB first assumed. Try it, but expect to
  write the NDK/JNI bridge from scratch — highest-effort candidate.
- `MobileMoE` (Meta AI) risk note (2026-09-14 research): paper-only.
  Real Samsung Galaxy S25 benchmarks exist in the arXiv paper
  (2605.27358), but no public model weights or code have been
  released anywhere. Not usable for a spike today — revisit if Meta
  later publishes a checkpoint.
- Lowest-risk starting point (2026-09-14 research): **Qwen2.5
  (1.5B–3B) GGUF, quantized (Q4_K_M ≈ 0.8–2 GB), run through
  `llama.cpp`'s existing Android bindings/example apps.** This is the
  most tested GGUF+Android combo, with working example projects
  already public — start the spike here.
- Backup if Qwen2.5 underperforms: Gemma 3 1B via Google's official
  `LiteRT-LM` (github.com/google-ai-edge/LiteRT-LM), which also has a
  sample Android app.
- General fallback: any 4B-class dense model quantized to about
  1–1.5 GB, still via `llama.cpp`'s Android bindings.
- Run whichever candidates actually build on a real Android phone.
  Measure real tokens/sec, not a desktop or Raspberry Pi number —
  those don't count.
- Only after a model is proven on-device does the real Chat tab work
  begin inside Raqm.

## Tier 2 — Bring-your-own-key (free)

- User pastes their own Anthropic or OpenAI API key into Settings.
- The key stays on the device only (secure local storage). It is
  never sent to a Raqm server.
- The app calls the AI provider directly from the phone.
- Zero cost and zero server work for Raqm. The user pays their own
  provider for their own usage.

## Tier 3 — Raqm-hosted AI credits (paid)

- Raqm runs a backend proxy. The proxy calls a commercial API
  (Anthropic or OpenAI) on the user's behalf.
- Each user gets a monthly credit allowance, tracked server-side.
- Raqm's own key never reaches the client — the proxy holds it.
- Self-hosting Raqm's own model instead of paying per API call is
  deferred until user volume makes that cheaper.

## Deferred — MCP data access (paid, later)

- Raqm exposes a user's financial data to their own external AI
  assistant (their own Claude, ChatGPT, etc.) over MCP.
- Needs an encrypted, server-side copy of the user's data and
  per-user login (OAuth-style), since a phone cannot stay reachable
  as a live server on its own.
- The most complex and most expensive tier to build safely. Comes
  after tiers 1–3 are shipped and working.

## Immediate next step

Tier 1 spike: stand up the throwaway test app with `llama.cpp`'s
Android bindings and Qwen2.5 (1.5B–3B) GGUF — the lowest-risk starting
point per the 2026-09-14 research. In parallel, attempt a `BitNet
b1.58` build for comparison. `MobileMoE` is paper-only and not usable
yet. Get a connected Android device and measure real tokens/sec.

**Comment by rahatsayyed** [comment]
## Research update — 2026-09-23

### Laya (github.com/NandhaKishorM/laya) — skip

Python 3.10+ only, no Android SDK, non-generative (classifier only). Cannot produce natural-language answers. Not usable for this use case at all.

### Gemini Nano — useful but device-gated

Gemini Nano runs via Android AICore, which is only available on specific high-end hardware: Pixel 8+, Galaxy S24+, and a handful of other flagships. It is NOT available on all Android devices. For Indian users (Raqm's target market), most devices are mid-range — Gemini Nano likely covers ~10–20% of real users at best.

**Gemini Nano is NOT heavier than Qwen2.5** — the opposite. It is a system model (no app download needed, Google manages it), so it has zero storage cost for the user. But device availability gates it, not size.

### Updated Tier 1 spike order

1. **Gemini Nano via Android AICore (ML Kit GenAI API)** — check at runtime whether the device supports it. If yes, use it: zero download, Google-managed, best experience on flagships.
2. **Qwen2.5 (1.5B–3B) GGUF via llama.cpp Android bindings** — fallback for all other devices (~3–4 GB RAM required, ~0.8–2 GB model download). Broader coverage.
3. **BitNet b1.58** — no official Android/NDK support, one failed Termux compile attempt (unresolved), no on-phone benchmarks anywhere. High effort, uncertain outcome. Low priority.
4. **MobileMoE** — paper-only (arXiv 2605.27358), no public weights or code. Not usable yet.

Best shipping strategy: detect AICore support at runtime, use Gemini Nano if available, otherwise fall back to llama.cpp + Qwen2.5. This gives flagship users a seamless experience and still works on mid-range hardware.

**Comment by rahatsayyed** [comment]
Qwen (3B) - Q3_K_M~1.59 MB~2.0 GBHighly Recommended. Cuts down file size significantly while keeping most of the 3B model's reasoning power.

**Comment by rahatsayyed** [comment]
## Architecture research update — 2026-09-23

### Fine-tuning a small model

Worth doing eventually, not yet. Fine-tuning improves intent parsing accuracy, Indian finance terminology (UPI, NACH, auto-debit), and reduces the system-prompt size needed on small context windows.

**Blocker: no user query data yet.** Fine-tuning with synthetic data often makes things worse. The right time is after shipping the base version and collecting real anonymized queries.

When ready: fine-tune base Qwen2.5 with LoRA/QLoRA on a GPU cloud (Modal, RunPod) → quantize to GGUF → ship via llama.cpp. ~1 day of work once training data exists.

---

### SLM clarification — two different things

**SLM as you described it (~300–400MB, one specific task):**
A task-specific edge model trained for one narrow job — intent classification, slot-filling, NER. Think DistilBERT-class or a fine-tuned Phi-1. These are NOT general assistants.

**SLM as the industry uses it (1–3B range):**
Just a smaller general LLM. Qwen2.5 1.5B, Gemma 3 1B, etc. These can do Q&A, reasoning, phrasing — but they are heavier (~0.8–2 GB).

Both are useful. They serve different roles.

---

### Recommended two-model architecture for Raqm

```
User question (free text)
        ↓
[Task model ~300MB] — parses intent into structured slot-fill
  e.g. type=spending_query, category=dining, period=august
        ↓
[Deterministic Kotlin/SQL] — runs real query, computes the number
  e.g. spent=12400, budget=15000, currency=INR
        ↓
[General LLM — Gemini Nano or Qwen2.5 1.5B] — phrases the answer
  e.g. "You spent Rs.12,400 on dining in August, Rs.2,600 under budget."
```

**Why this works well:**
- Task model fires on every query — cheap, fast, battery-friendly
- General LLM only fires to phrase the final answer — expensive but infrequent
- AI never invents a number — all figures come from SQL
- Task model can be fine-tuned on real user queries later without touching the LLM

**Task model candidates (Android-compatible):**
- Rule-based (0 MB) — valid v1 option: regex + keyword matching covers most finance query patterns
- DistilBERT (~67 MB) — standard intent classification, ONNX runtime on Android
- TinyBERT (~55 MB) — even smaller, similar quality
- Fine-tuned Phi-1 (~1.3 GB quantized) — overkill for pure classification, but supports structured output

**v1 spike recommendation:**
Start with rule-based intent routing (Kotlin, ~20 lines) + Gemini Nano or Qwen2.5 for phrasing. Upgrade the routing to a real task model once real query patterns are known from actual users.

**Comment by rahatsayyed** [comment]
## Research update — finance-specific SLM feasibility (2026-09-23)

### Does a pre-built finance-specific small model exist?

**No.** Every existing finance-domain LLM (AdaptLLM/finance-LLM, FinGPT, finance-chat) is 7B+ parameters — minimum ~4GB quantized. Not mobile-runnable. FinBERT is encoder-only (classifies sentiment, cannot generate text). There is no sub-1GB personal finance Q&A model anywhere today.

**The architecture the user described is valid — it just needs to be built.**

---

### Confirmed architecture

```
User question (free text)
        ↓
[Finance SLM ~349MB] — Qwen2.5-0.5B fine-tuned on finance intents
  outputs: {"intent": "spending_query", "category": "dining", "period": "august"}
  (grammar-constrained decoding enforces valid JSON as hard fallback)
        ↓
[Deterministic Kotlin/SQL] — runs real query, computes actual numbers
  outputs: {"spent": 12400, "budget": 15000, "currency": "INR"}
        ↓
[Phrase layer] — same model second pass, or Gemini Nano
  outputs: "You spent Rs.12,400 on dining in August, Rs.2,600 under budget."
```

This is exactly what the user proposed: a small model that only understands finance, returns JSON, with separate logic to compute real numbers and phrase the result.

---

### Base model for fine-tuning

**Qwen/Qwen2.5-0.5B-Instruct** (https://huggingface.co/Qwen/Qwen2.5-0.5B-Instruct)

- GGUF at IQ4_XS: ~349MB (https://huggingface.co/bartowski/Qwen2.5-0.5B-Instruct-GGUF)
- Android: runs via llama.rn (React Native / Expo binding for llama.cpp)
- Structured output: native JSON support in Qwen2.5 + grammar-constrained decoding via llama.cpp GBNF as fallback
- Fine-tuning: supervised fine-tuning on {question → intent JSON} pairs, then re-quantize to GGUF

Runner-up if even smaller is needed: **SmolLM2-360M-Instruct** (~200MB Q4) — smaller but needs more fine-tuning effort for reliable JSON output on narrow schemas.

---

### FinBERT / FinGPT — confirmed dead ends for this use case

- **FinBERT**: encoder-only, classifies sentiment (positive/neutral/negative), cannot generate text, no GGUF
- **FinGPT**: fine-tuning framework on 7B+ base models, focused on stock/market analysis — not personal budgeting, not mobile-runnable
- **finance-LLM (AdaptLLM)**: 7B, ~4GB minimum quantized — too large

---

### What fine-tuning requires

Training data: curated pairs of personal finance questions mapped to intent JSON. Example:
- "how much did I spend on food last month?" → `{"intent": "spending_query", "category": "food", "period": "last_month"}`
- "where should I cut to save Rs.5000?" → `{"intent": "advisor", "goal_amount": 5000}`
- "what is my dining budget?" → `{"intent": "budget_lookup", "category": "dining"}`

No user data exists yet — build synthetic training data first, fine-tune with LoRA/QLoRA on a GPU cloud (Modal, RunPod), re-quantize to IQ4_XS GGUF, ship via llama.rn.

Timing: fine-tune after the on-device spike proves the runtime works. Do not fine-tune before a base model is running on a real device.
